File: lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode A, TreeNode B) {
        if(root == null) return null;
        if(A.val == root.val && B.val == root.val) return root;

        if(A.val == root.val) return root;
        if(B.val == root.val) return root;

        TreeNode leftAns = lowestCommonAncestor(root.left,A,B);
        TreeNode rightAns = lowestCommonAncestor(root.right,A,B);

        if(leftAns != null && rightAns != null) return root;
        else if(leftAns != null) return leftAns;
        else return rightAns;
    }
}

File: lowest-common-ancestor-of-a-binary-tree/lowest-common-ancestor-of-a-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(root == NULL) return root;
        if(root->val == p->val or root->val == q->val) return root;
        
        TreeNode* leftAns = lowestCommonAncestor(root->left,p,q);
        TreeNode* rightAns = lowestCommonAncestor(root->right,p,q);
        
        if(leftAns and rightAns) return root;
        else if(leftAns and !rightAns) return leftAns;
        else if(!leftAns and rightAns) return rightAns;
        else return NULL;
    }
};